Initial commit of the Cds template

rakpar opened the pull request:
This is just an initial PR. I expect there to be comments since I am assuming several things on the code. I also have TODOs which need to be fixed.

Major question: how do I get the template tagged as experimentation since we are referring to our preview nuget as well.

dnfclas commented:
@__rakpar__,
Thanks for your contribution.
To ensure that the project team has proper rights to use your work, please complete the Contribution License Agreement at https://cla2.dotnetfoundation.org.

It will cover your contributions to all .NET Foundation-managed open source projects.
_Thanks,
.NET Foundation Pull Request Bot_

MikeStall commented on Templates/CDS-CSharp/SecurityHelper.csx:
```diff
@@ -0,0 +1,32 @@
+            }
+        };
+
+        return settings.CreateClient(requestMessage, logger as TraceWriterTelemetryBridge ?? new TraceWriterTelemetryBridge());
```
What information are we pulling from the request?

rakpar replied:
See the EnvironmentValues class in run.csx. We eventually want to replace it with the auth info that we will get from the ambient context.[https://github.com/rakpar/azure-webjobs-sdk-templates/blob/rakpar-cds/Templates/CDS-CSharp/run.csx](url)

MikeStall replied:
What does "ambient context" mean in this case? The EnvVar look like appsetting values, not per-request values. 

---
In reply to: [110759417](https://github.com/Azure/azure-webjobs-sdk-templates/pull/423#discussion_r110759417) [](ancestors = 110759417)

dnfclas commented:
__@rakpar__, thanks for signing the contribution license agreement. We will now validate the agreement and then the pull request.
<br />_Thanks, .NET Foundation Pull Request Bot_

MikeStall commented on Templates/CDS-CSharp/run.csx:
```diff
@@ -0,0 +1,67 @@
+using System.Net.Http;
+using System.Net.Http.Headers;
+
+class EnvironmentValues
```
>EnvironmentValues [](start = 6, length = 17)

Should these eventually be appsettings?

rakpar replied:
I believe there are discussions going on about the design for this. But yes, appsettings is one of the probable places. Nimak and Matthew Henderson are the PMs on this.

nimakms replied:
Agreed these will eventually be app settings driven by Azure function UI to setup EasyAuth and Environment selection. Consensus with Matthew was for the first iteration to ship with these hardcoded.

nimakms commented on Templates/CDS-CSharp/run.csx:
```diff
@@ -0,0 +1,67 @@
+        // TODO: pick this from req parameters
+        var query = client.GetRelationalEntitySet<ProductCategory>()
+            .CreateQueryBuilder()
+            .Where(pc => pc.Name == "Surface")
```
You need   || pc.Name == "Phone" here too, to clean up everything.

rakpar replied:
I had purposely removed it to just get the Surface from req parameters but noticed the rest of the code needs it. Adding it back. Will take care of the req params some time later

nimakms commented on Templates/CDS-CSharp/run.csx:
```diff
@@ -0,0 +1,66 @@
+﻿#load ".\SecurityHelper.csx"
```
Sorry for creeping (agile) spec but we need the following as a comment here to cover our legal requirements for SDK.

// The contents of this template are subject to license terms of the Microsoft Common 
// Data Service Software Development Kit (https://go.microsoft.com/fwlink/?linkid=842862).

soninaren commented on Templates/CDS-CSharp/metadata.json:
```diff
@@ -0,0 +1,15 @@
+    "description": "$cdsWebHookCSharp_description",
+    "defaultFunctionName": "CdsWebHookCSharp",
+    "language": "C#",
+    "category": [
```
Change the value of category to `[ "$temp_category_experimental" ]`. This will get the template tagged as experimental.

rakpar commented:
I will be waiting for @MikeStall to push a change where we can have the Cds binding in place. I will modify the template accordingly then.

soninaren commented:
Please test the template using the steps in here
https://github.com/Azure/azure-webjobs-sdk-templates/#testing-templates

rakpar commented:
I have tested the template and it works fine with the steps outlined at https://github.com/Azure/azure-webjobs-sdk-templates/#testing-templates. The first phase of this is ready to be signed off. @nimakms, @soninaren, @MikeStall, please have a look here.

nimakms commented on Templates/CDS-CSharp/run.csx:
```diff
@@ -0,0 +1,68 @@
+
+class EnvironmentValues
+{
+    internal const string Tenant = "[[Replace with AAD tenant for CDS value]]";
```
Assuming we can work on removing tenant in a future revision. also that probably requires changes in the SDK right? How should we track that work.

nimakms commented on Templates/CDS-CSharp/run.csx:
```diff
@@ -0,0 +1,68 @@
+{
+    internal const string Tenant = "[[Replace with AAD tenant for CDS value]]";
+    internal const string EnvironmentId = "[[Replace with PowerApps environment ID value]]";
+    internal const string ApplicationId = "[[Replace with Function application ID value]]";
```
Are we getting these values from EasyAuth later? If yes, are we tracking in another work item?

nimakms commented on Documentation/cdsIn.md:
```diff
@@ -0,0 +1,11 @@
+﻿# CdsWebHook - C<span>#</span>
+
+The `CdsWebHook` makes it incredibly easy to have your CDS functions executed via an HTTP call to your function.
```
Is WebHook the right term? I'm a bit ignorant about web-hooks, wo please educate me what they are, and if this template is in fact a web-hook.

nimakms commented on Templates/CDS-CSharp/function.json:
```diff
@@ -0,0 +1,16 @@
+﻿{
```
@mattchenderson - You mentioned there is a setting that enables the EasyAuth warning on template function creation. What is that exactly, and where does that go?

mattchenderson commented:
To enable the EasyAuth config warning, add the following to metadata.json

`
   "warning": {
        "type": "Warning",
        "text": "$ADDToken_warningNotCongigured",
        "addLinkToAuth": true
    }
`

You'll likely want to pick a specific string of your own for the text. For en-us, my recommendation would be:
"This template requires an AAD configuration for the function app. The template needs the following permission:
- Common Data Service (Access the Common Data Service)"
I'll leave the specifics of the permission to the CDS team. The idea is to have the pattern "ResourceName (PermissionName)" for permissions in such a list.

We have a goal of eventually having the template just provide the permission list, and we'll take care of the rest, but that will not be part of this initial release.

mattchenderson commented:
The key being used here should not be aadToken_warningNotConfigured. That will be used by other components, and it shouldn't include CDS-specific permissions. I’d propose a new key named cdsTemplate_warningNotConfigured or cdsWebHook_warningNotConfigured instead.

rakpar commented:
@Matthew Henderson<mailto:Matthew.Henderson@microsoft.com>, just addressed all the comments. Could you help me approve and merge the PR?

From: Matthew Henderson [mailto:notifications@github.com]
Sent: Wednesday, April 26, 2017 11:31 AM
To: Azure/azure-webjobs-sdk-templates <azure-webjobs-sdk-templates@noreply.github.com>
Cc: Rakesh Parida <Rakesh.Parida@microsoft.com>; Mention <mention@noreply.github.com>
Subject: Re: [Azure/azure-webjobs-sdk-templates] Initial commit of the Cds template (#423)


The key being used here should not be aadToken_warningNotConfigured. That will be used by other components, and it shouldn't include CDS-specific permissions. I’d propose a new key named cdsTemplate_warningNotConfigured or cdsWebHook_warningNotConfigured instead.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FAzure%2Fazure-webjobs-sdk-templates%2Fpull%2F423%23issuecomment-297500979&data=02%7C01%7CRakesh.Parida%40microsoft.com%7C20438ba4e7bb452b506908d48cd26e4f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636288282811437808&sdata=I1v3Ku2yvkFFe%2FWYDpVtwb6mS1RxHBHLvrHE67CzlCY%3D&reserved=0>, or mute the thread<https://na01.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAY8S2AwmV5wkAHVo_lVqiVI2KVAiTgQ3ks5rz410gaJpZM4M5R_8&data=02%7C01%7CRakesh.Parida%40microsoft.com%7C20438ba4e7bb452b506908d48cd26e4f%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C636288282811437808&sdata=lMaxOPzwzg%2FLgK8s9yA6x0ossyuinimj2NZ5fuxzq%2Bk%3D&reserved=0>.

mattchenderson commented on Resources/Resources.resx:
```diff
@@ -1,1161 +1,1170 @@
+    
+    mimetype: application/x-microsoft.net.object.soap.base64
+    value   : The object must be serialized with 
+       
```
"PowerApps Runtime Service - Access PowerApps Runtime Service"

rakpar replied:
Resolved in a new update

soninaren commented on Templates/CDS-CSharp/metadata.json:
```diff
@@ -0,0 +1,19 @@
+    "userPrompt": [
+        "connection",
+        "path"
+    ],
```
change this to 
```JSON
"userPrompt": [
        "authLevel"
    ]

```
